### DISCUSSION TITLE
How do i Use the New-PesterConfiguration ? I can't get it to set Verbosity to Diagnostic in my output

### DISCUSSION THREAD

**Discussion by es-cs1** [pester/Pester#2328] [Q&A] [answered]
from https://pester.dev/docs/usage/configuration#pesterpreference
I have copied the following code into a file:

```powershell

$PesterPreference = [PesterConfiguration]::Default
$PesterPreference.Debug.WriteDebugMessages = $true
$PesterPreference.Debug.WriteDebugMessagesFrom = "Mock"

BeforeAll {
    function a { "hello" }
}
Describe "pester preference" {
    It "mocks" {
        Mock a { "mock" }
        a | Should -Be "mock"
    }
}

```
I then invoke it 
Invoke-Pester .\mytest.ps1

I then get
Starting discovery in 1 files.
Discovery found 1 tests in 94ms.
Running tests.
[+] C:\Users\<my_user>\my_folder\mytest.ps1 907ms (630ms|185ms)
Tests completed in 909ms
Tests Passed: 1, Failed: 0, Skipped: 0 NotRun: 0

Why arent the debug outputs showing?

EDIT: I found if I _DIRECTLY_ execute the script e.g.  `.\myscript.ps1`  the debug output shows... but it is hidden if i use `Invoke-Pester` command

I want to in general use the configuration detailed in https://pester.dev/docs/usage/output, but it wont work.

heres another test
my test file:

```powershell

    Describe "pester preference" {
        It "test 1" {
        
         5 | Should -Be 5
       }

    It "test2 " {
        5 | Should -Be 5
        5 | Should -Be 5
        5 | Should -Be 5
    }

    It "test3 " {
        5 | Should -Be 5
        5 | Should -Be 5
        5 | Should -Be 5
    }
}
```

**Then I set the configuration... Am I doing it wrong somehow?**

```
$conf = New-PesterConfiguration
$conf.Output.Verbosity = "Diagnostic"
Invoke-Pester .\mytest2.ps1
```

Starting discovery in 1 files.
Discovery found 3 tests in 88ms.
Running tests.
[+] C:\Users\<my_user>\my_folder\mytest\mytest2.ps1 345ms (106ms|152ms)
Tests completed in 348ms
Tests Passed: 3, Failed: 0, Skipped: 0 NotRun: 0

I would have expected the Diagnostic Level verbosity to show each test , and maybe the "Assert statements" as they execute?

My Get-Module -ListAvailable shows

```
Script     5.4.0                 Pester                              Desk      {Invoke-Pester, Describe, Context, It…}
Script     3.4.0                 Pester                              Desk      {Describe, Context, It, Should…}
```

**Comment by fflaten** [comment, marked as answer]
Thanks for reporting this!

Right above the first sample it says _"while running interactively"_. Interactive execution means running the script directly without calling `Invoke-Pester`, typically by pressing <kbd>F5</kbd> in VSCode or ISE. So it's intended behavior, but poorly explained and something we should update 🙂 

`$PesterPreference` is session preferences (like `$PSDefaultParameterValues` in PowerShell) and any values has to be set before calling `Invoke-Pester` which merges it with any provided parameters. In general I'd recommend to avoid `$PesterPreference` unless you want to include a few settings in your profile

Instead you could use a `test.ps1` file where you define your configuration and pass it to `Invoke-Pester -Configuration $conf`. Maybe even post-process the results object (returned by setting `$conf.Run.PassThru = $true`).

```ps1
$conf = New-PesterConfiguration
$conf.Output.Verbosity = "Diagnostic"
$conf.Run.Path = "$PSScriptRoot\mytest2.ps1" # or "$PSScriptRoot/tst" folder ++
Invoke-Pester -Configuration $conf
```

**Reply by es-cs1** [reply]
Thank you! I love the new documentation website, keep up the great work.
